File: jiuwenswarm/channels/ide/packages/jetbrains-plugin/src/main/kotlin/com/jiuwenswarm/plugin/swarm/SwarmStateManager.kt
package com.jiuwenswarm.plugin.swarm

import com.google.gson.Gson
import com.google.gson.JsonObject
import com.intellij.openapi.diagnostic.logger
import java.io.File

private val LOG = logger<SwarmStateManager>()
private val gson = Gson()

/**
 * Maintains in-memory swarm state built from team event deltas that arrive
 * over the WebSocket as "team.event:{json}" prefixed chat.delta messages.
 *
 * All public methods are @Synchronized — onJiuwenMessage runs on a background
 * thread while postSnapshot reads may happen on any thread.
 */
class SwarmStateManager {

    private var sessionId: String = ""
    private var teamName: String = ""
    private val lanes = LinkedHashMap<String, AgentLane>()   // insertion order = join order
    private val tasks = LinkedHashMap<String, TeamTask>()
    private val messages = ArrayDeque<TeamMessage>()          // ring buffer, max 50
    @Volatile private var lastEventAt: Long = 0L

    // ──────────────────────────────────────────
    // Public API
    // ──────────────────────────────────────────

    /**
     * Parse the JSON string that follows "team.event:" and apply it to state.
     * Expected shape: { "event": { "type": "team.member.spawned", ... } }
     */
    @Synchronized
    fun applyTeamEvent(json: String) {
        runCatching {
            val root = gson.fromJson(json, JsonObject::class.java)
            val event = normalizeEvent(root.getAsJsonObject("event") ?: root)  // tolerate flat shape
            val type = event.get("type")?.asString ?: return
            lastEventAt = event.get("timestamp")?.asLong ?: System.currentTimeMillis()

            // Capture team name from the first event that carries it
            event.get("team_name")?.asString?.let { if (teamName.isEmpty()) teamName = it }

            when {
                type == "team.member.spawned"          -> onMemberSpawned(event)
                type == "team.member.status_changed"   -> onMemberStatusChanged(event)
                type == "team.member.execution_changed"-> onMemberExecutionChanged(event)
                type == "team.member.activity_changed" -> onMemberActivityChanged(event)
                type == "team.member.shutdown"         -> onMemberShutdown(event)
                type == "team.task.created"            -> onTaskCreated(event)
                type == "team.task.claimed"            -> onTaskClaimed(event)
                type == "team.task.started"            -> onTaskStarted(event)
                type == "team.task.completed"          -> onTaskCompleted(event)
                type == "team.task.cancelled"          -> onTaskCancelled(event)
                type == "team.task.status_snapshot"    -> onTaskStatusSnapshot(event)
                type.startsWith("team.message.")       -> onMessage(event)
                else -> LOG.debug("SWARM → unhandled event type: $type")
            }
        }.onFailure { LOG.warn("SWARM → failed to apply team event: ${it.message}") }
    }

    /**
     * Map the gateway's team-event field names to the shape this manager consumes.
     * The gateway sends member_id/name/mode/team_id; the lanes here use
     * member_name/display_name/role/team_name. Runs before dispatch so every
     * handler below reads one consistent vocabulary.
     */
    private fun normalizeEvent(raw: JsonObject): JsonObject {
        val e = raw.deepCopy()
        if (e.has("member_id") && !e.has("member_name")) {
            e.get("member_id")?.takeIf { it.isJsonPrimitive }?.asString?.let { e.addProperty("member_name", it) }
        }
        if (e.has("name") && !e.has("display_name")) {
            e.get("name")?.takeIf { it.isJsonPrimitive }?.asString?.let { e.addProperty("display_name", it) }
        }
        if (e.has("mode") && !e.has("role")) {
            val mode = e.get("mode")
            if (mode != null && mode.isJsonPrimitive) {
                e.addProperty("role", mode.asString.uppercase())
            } else if (mode != null) {
                e.add("role", mode)
            }
        }
        if (e.has("team_id") && !e.has("team_name")) {
            e.get("team_id")?.takeIf { it.isJsonPrimitive }?.asString?.let { e.addProperty("team_name", it) }
        }
        return e
    }

    /** Map a gateway status into the map's lane vocabulary (READY/BUSY/PAUSED/SHUTDOWN). */
    private fun normalizeLaneStatus(s: String?): String? {
        if (s.isNullOrEmpty()) return null
        return when (s.uppercase()) {
            "IDLE"      -> "READY"
            "RUNNING"   -> "RUNNING"
            "COMPLETED" -> "SHUTDOWN"
            else        -> s.uppercase()
        }
    }

    /**
     * Called from the chat.tool_call / chat.tool_update branches of onJiuwenMessage.
     * memberName may be null if the server does not include it in the payload.
     * args (the parsed tool arguments) lets the feed say what an action touches —
     * e.g. a file_path, a glob pattern, or the shell command.
     */
    @Synchronized
    fun applyToolCall(toolName: String, filePath: String?, memberName: String?, args: JsonObject? = null) {
        val name = memberName ?: return   // cannot attribute without a member name
        val lane = lanes.getOrPut(name) { stubLane(name) }
        lane.lastToolName = toolName
        val activity = formatActivity(toolName, filePath, args)
        lane.currentActivity = activity
        lane.lastActivePath = filePath
        lane.lastActiveAt = System.currentTimeMillis()
        lane.status = "BUSY"
        pushFeed(lane, activity, lane.lastActiveAt)
        lastEventAt = lane.lastActiveAt
    }

    /** Model call started — show "thinking…" on the lane instead of drifting into idle. */
    @Synchronized
    fun applyModelCallStart(memberName: String?) {
        val name = memberName ?: return
        val lane = lanes.getOrPut(name) { stubLane(name) }
        lane.status = "BUSY"
        lane.currentActivity = "thinking…"
        lane.lastActiveAt = System.currentTimeMillis()
        pushFeed(lane, "thinking…", lane.lastActiveAt)
        lastEventAt = lane.lastActiveAt
    }

    /** First streamed token for a call: flip the lane from "thinking…" to "generating…". */
    @Synchronized
    fun applyModelTokenStart(memberName: String?) {
        val name = memberName ?: return
        val lane = lanes[name] ?: return
        if (lane.currentActivity != "thinking…") return
        lane.currentActivity = "generating…"
        lane.lastActiveAt = System.currentTimeMillis()
    }

    @Synchronized
    fun applyModelCallEnd(memberName: String?) {
        val name = memberName ?: return
        val lane = lanes[name] ?: return
        if (lane.currentActivity == "thinking…" || lane.currentActivity == "generating…") {
            lane.currentActivity = null
        }
        lane.lastActiveAt = System.currentTimeMillis()
    }

    @Synchronized
    fun snapshot(): SwarmSnapshot {
        // Stable insertion order (join order) — lanes must not jump around as
        // statuses change; the webview renders status separately from position.
        val lanesList = lanes.values.toList()
        val sortedTasks = tasks.values.sortedWith(compareBy { it.createdAt })
        return SwarmSnapshot(
            sessionId = sessionId,
            teamName = teamName,
            lanes = lanesList,
            tasks = sortedTasks,
            messages = messages.toList(),
            lastEventAt = lastEventAt,
        )
    }

    @Synchronized
    fun reset(newSessionId: String) {
        sessionId = newSessionId
        teamName = ""
        lanes.clear()
        tasks.clear()
        messages.clear()
        lastEventAt = 0L
    }

    fun isEmpty(): Boolean = lanes.isEmpty()

    // ──────────────────────────────────────────
    // Event handlers
    // ──────────────────────────────────────────

    private fun onMemberSpawned(e: JsonObject) {
        val name = e.get("member_name")?.asString ?: return
        val status = normalizeLaneStatus(e.get("status")?.asString) ?: "READY"
        val prompt = e.get("prompt")?.asString
        val existing = lanes[name]
        if (existing != null) {
            // Update display info in case of restart
            existing.displayName = e.get("display_name")?.asString ?: existing.displayName
            existing.role = e.get("role")?.asString ?: existing.role
            existing.status = status
            if (!prompt.isNullOrEmpty()) {
                existing.currentTaskTitle = prompt
                existing.currentActivity = "starting"
            }
            existing.startedAt = existing.startedAt ?: lastEventAt
            pushFeed(existing, "spawned", lastEventAt)
            existing.lastActiveAt = lastEventAt
        } else {
            lanes[name] = AgentLane(
                memberName = name,
                displayName = e.get("display_name")?.asString ?: name,
                role = e.get("role")?.asString ?: "TEAMMATE",
                status = status,
                currentTaskTitle = prompt,
                currentActivity = if (prompt.isNullOrEmpty()) null else "starting",
                lastActiveAt = lastEventAt,
                startedAt = lastEventAt,
                activityFeed = mutableListOf(LaneFeedEntry(text = "spawned", at = lastEventAt)),
            )
        }
    }

    private fun onMemberStatusChanged(e: JsonObject) {
        val name = e.get("member_name")?.asString ?: return
        val lane = lanes.getOrPut(name) { stubLane(name) }
        val newStatus = e.get("new_status")?.takeIf { it.isJsonPrimitive }?.asString
            ?: e.get("status")?.takeIf { it.isJsonPrimitive }?.asString
        val status = normalizeLaneStatus(newStatus)
        status?.let { lane.status = it }
        val outcome = e.get("outcome")?.asString
        if (!outcome.isNullOrEmpty() && status == "SHUTDOWN") {
            lane.currentActivity = "done · $outcome"
        }
        if (status == "SHUTDOWN") pushFeed(lane, "finished", lastEventAt)
        else if (status == "PAUSED") pushFeed(lane, "paused", lastEventAt)
        lane.lastActiveAt = lastEventAt
    }

    private fun onMemberExecutionChanged(e: JsonObject) {
        val name = e.get("member_name")?.asString ?: return
        val lane = lanes.getOrPut(name) { stubLane(name) }
        val execution = e.get("new_status")?.takeIf { it.isJsonPrimitive }?.asString
            ?: e.get("execution_status")?.takeIf { it.isJsonPrimitive }?.asString
        if (!execution.isNullOrEmpty()) lane.executionStatus = execution.uppercase()
        lane.lastActiveAt = lastEventAt
    }

    /** Live mid-run activity (e.g. a swarmflow worker tool call) → lane activity. */
    private fun onMemberActivityChanged(e: JsonObject) {
        val name = e.get("member_name")?.asString ?: return
        val activity = e.get("activity")?.asString ?: return
        val lane = lanes.getOrPut(name) { stubLane(name) }
        lane.status = "BUSY"
        lane.currentActivity = activity
        pushFeed(lane, activity, lastEventAt)
        lane.lastActiveAt = lastEventAt
    }

    private fun onMemberShutdown(e: JsonObject) {
        val name = e.get("member_name")?.asString ?: return
        val lane = lanes[name] ?: return
        lane.status = "SHUTDOWN"
        lane.executionStatus = "IDLE"
        lane.currentActivity = null
        pushFeed(lane, "shutdown", lastEventAt)
        lane.lastActiveAt = lastEventAt
    }

    private fun onTaskCreated(e: JsonObject) {
        val id = e.get("task_id")?.asString ?: return
        tasks[id] = TeamTask(
            taskId = id,
            title = e.get("title")?.asString ?: id,
            status = "pending",
            createdAt = lastEventAt,
        )
    }

    private fun onTaskClaimed(e: JsonObject) {
        val id = e.get("task_id")?.asString ?: return
        val task = tasks.getOrPut(id) { TeamTask(id, e.get("title")?.asString ?: id, "pending") }
        // Gateway names the claimer member_id; normalizeEvent maps it to member_name.
        task.assignee = e.get("assignee")?.asString ?: e.get("member_name")?.asString
        // Server carries the authoritative status (e.g. "in_progress") — prefer it;
        // fall back to started semantics for older gateways that only mark claimed.
        task.status = e.get("status")?.asString ?: "in_progress"
        // Link task into the lane so the board reflects the current work.
        task.assignee?.let { memberName ->
            lanes[memberName]?.let { lane ->
                lane.currentTaskId = id
                lane.currentTaskTitle = task.title
                lane.lastActiveAt = lastEventAt
            }
        }
    }

    /** Convergence event: apply a task's authoritative status/assignee snapshot. */
    private fun onTaskStatusSnapshot(e: JsonObject) {
        val id = e.get("task_id")?.asString ?: return
        val status = e.get("status")?.asString
        val assignee = e.get("assignee")?.asString ?: e.get("member_name")?.asString
        val task = tasks.getOrPut(id) { TeamTask(id, e.get("title")?.asString ?: id, "pending") }
        if (!status.isNullOrEmpty()) task.status = status
        if (!assignee.isNullOrEmpty()) task.assignee = assignee
        task.assignee?.let { memberName ->
            lanes[memberName]?.let { lane ->
                if (task.status == "completed" && lane.currentTaskId == id) {
                    lane.currentTaskId = null
                    lane.currentTaskTitle = null
                    lane.tasksDone++
                    lane.lastActiveAt = lastEventAt
                }
            }
        }
    }

    private fun onTaskStarted(e: JsonObject) {
        val id = e.get("task_id")?.asString ?: return
        val task = tasks.getOrPut(id) { TeamTask(id, id, "pending") }
        task.status = "in_progress"
        task.assignee = e.get("assignee")?.asString ?: task.assignee
        // Link task into the lane
        task.assignee?.let { memberName ->
            lanes[memberName]?.let { lane ->
                lane.currentTaskId = id
                lane.currentTaskTitle = task.title
                lane.lastActiveAt = lastEventAt
            }
        }
    }

    private fun onTaskCompleted(e: JsonObject) {
        val id = e.get("task_id")?.asString ?: return
        val task = tasks[id] ?: return
        task.status = "completed"
        // Clear task from lane
        task.assignee?.let { memberName ->
            lanes[memberName]?.let { lane ->
                if (lane.currentTaskId == id) {
                    lane.currentTaskId = null
                    lane.currentTaskTitle = null
                    lane.tasksDone++
                    lane.lastActiveAt = lastEventAt
                }
            }
        }
    }

    private fun onTaskCancelled(e: JsonObject) {
        val id = e.get("task_id")?.asString ?: return
        val task = tasks[id] ?: return
        task.status = "cancelled"
        task.assignee?.let { memberName ->
            lanes[memberName]?.let { lane ->
                if (lane.currentTaskId == id) {
                    lane.currentTaskId = null
                    lane.currentTaskTitle = null
                    lane.lastActiveAt = lastEventAt
                }
            }
        }
    }

    private fun onMessage(e: JsonObject) {
        val from = e.get("from_member")?.asString ?: return
        val lane = lanes.getOrPut(from) { stubLane(from) }
        lane.messageCount++
        lane.lastActiveAt = lastEventAt
        // Capture message content for the inter-agent message log
        val content = e.get("content")?.asString ?: return
        val to = e.get("to_member")?.asString
        messages.addLast(TeamMessage(from = from, to = to, content = content, timestamp = lastEventAt))
        if (messages.size > 50) messages.removeFirst()
    }

    // ──────────────────────────────────────────
    // Helpers
    // ──────────────────────────────────────────

    private fun stubLane(memberName: String) = AgentLane(
        memberName = memberName,
        displayName = memberName,
        role = "TEAMMATE",
        lastActiveAt = lastEventAt,
        startedAt = lastEventAt.takeIf { it != 0L } ?: System.currentTimeMillis(),
    )

    /** Append one entry to a lane's activity feed, keeping it capped.
     *  A duplicate of the previous entry within 1s (the tool_call + tool_update
     *  pair for one call) is collapsed into the existing row instead of flooding
     *  the feed; genuinely separate calls still get their own rows. */
    private fun pushFeed(lane: AgentLane, text: String, at: Long) {
        val last = lane.activityFeed.lastOrNull()
        if (last != null && last.text == text && at - last.at < 1000) {
            last.at = at
            return
        }
        lane.activityFeed.add(LaneFeedEntry(text = text, at = at))
        while (lane.activityFeed.size > 10) lane.activityFeed.removeAt(0)
    }

    private fun formatActivity(toolName: String, filePath: String?, args: JsonObject?): String {
        val base = filePath?.let { File(it).name } ?: filePath
        val pattern = args?.get("pattern")?.takeIf { it.isJsonPrimitive }?.asString
        val command = args?.get("command")?.takeIf { it.isJsonPrimitive }?.asString
            ?: args?.get("cmd")?.takeIf { it.isJsonPrimitive }?.asString
            ?: args?.get("command_line")?.takeIf { it.isJsonPrimitive }?.asString
        val taskId = args?.get("task_id")?.takeIf { it.isJsonPrimitive }?.asString
        val to = args?.get("to")?.takeIf { it.isJsonPrimitive }?.asString
        return when (toolName) {
            "write_file", "create_file" -> if (base != null) "writing · $base" else "writing"
            "str_replace_editor", "edit_file" -> if (base != null) "editing · $base" else "editing"
            "read_file"                 -> if (base != null) "reading · $base" else "reading"
            "bash", "run_command", "powershell", "shell", "cmd"
                                        -> if (command != null) "running · ${command.take(40)}" else "running command"
            "list_files"                -> if (pattern != null) "listing · $pattern" else if (base != null) "listing · $base" else "listing"
            "glob", "grep", "search_files"
                                        -> if (pattern != null) "searching · $pattern" else if (base != null) "searching · $base" else "searching"
            "view_task"                 -> if (taskId != null) "viewing task · $taskId" else "viewing task"
            "claim_task"                -> if (taskId != null) "claiming task · $taskId" else "claiming task"
            "send_message"              -> if (to != null) "messaging · $to" else "messaging"
            else                        -> toolName
        }
    }
}

File: jiuwenswarm/channels/ide/packages/vscode-extension/src/swarm/SwarmStateManager.ts
import * as path from 'path';
import { AgentLane, TeamTask, TeamMessage, SwarmSnapshot } from './SwarmState';

/**
 * Mirrors SwarmStateManager.kt — maintains in-memory swarm state built from
 * team event deltas that arrive over the WebSocket as "team.event:{json}"
 * prefixed chat.delta messages.
 */
export class SwarmStateManager {
  private sessionId = '';
  private teamName = '';
  // Map preserves insertion order so join order is maintained
  private lanes = new Map<string, AgentLane>();
  private tasks = new Map<string, TeamTask>();
  private messages: TeamMessage[] = [];  // ring buffer, max 50
  private lastEventAt = 0;

  // ──────────────────────────────────────────
  // Public API
  // ──────────────────────────────────────────

  /**
   * Parse the JSON string that follows "team.event:" and apply it to state.
   * Expected shape: { "event": { "type": "team.member.spawned", ... } }
   */
  applyTeamEvent(json: string): void {
    try {
      const root = JSON.parse(json) as Record<string, unknown>;
      const event = this.normalizeEvent(((root['event'] as Record<string, unknown>) ?? root));
      const type = event['type'] as string | undefined;
      if (!type) return;
      this.lastEventAt = (event['timestamp'] as number | undefined) ?? Date.now();
      const teamName = event['team_name'] as string | undefined;
      if (teamName && !this.teamName) this.teamName = teamName;

      switch (type) {
        case 'team.member.spawned':           this.onMemberSpawned(event); break;
        case 'team.member.status_changed':    this.onMemberStatusChanged(event); break;
        case 'team.member.execution_changed': this.onMemberExecutionChanged(event); break;
        case 'team.member.activity_changed':  this.onMemberActivityChanged(event); break;
        case 'team.member.shutdown':          this.onMemberShutdown(event); break;
        case 'team.task.created':             this.onTaskCreated(event); break;
        case 'team.task.claimed':             this.onTaskClaimed(event); break;
        case 'team.task.started':             this.onTaskStarted(event); break;
        case 'team.task.completed':           this.onTaskCompleted(event); break;
        case 'team.task.cancelled':           this.onTaskCancelled(event); break;
        case 'team.task.status_snapshot':     this.onTaskStatusSnapshot(event); break;
        default:
          if (type.startsWith('team.message.')) this.onTeamMessage(event);
      }
    } catch {
      // malformed JSON — ignore silently
    }
  }

  /**
   * Map the gateway's team-event field names to the shape this manager consumes.
   * The gateway sends member_id/name/mode/team_id; the lanes here use
   * member_name/display_name/role/team_name. Runs before dispatch so every
   * handler below reads one consistent vocabulary.
   */
  private normalizeEvent(raw: Record<string, unknown>): Record<string, unknown> {
    const e: Record<string, unknown> = { ...raw };
    if (e['member_id'] !== undefined && e['member_name'] === undefined) e['member_name'] = e['member_id'];
    if (e['name'] !== undefined && e['display_name'] === undefined) e['display_name'] = e['name'];
    if (e['mode'] !== undefined && e['role'] === undefined) {
      e['role'] = typeof e['mode'] === 'string' ? e['mode'].toUpperCase() : e['mode'];
    }
    if (e['team_id'] !== undefined && e['team_name'] === undefined) e['team_name'] = e['team_id'];
    return e;
  }

  /** Map a gateway status into the map's lane vocabulary (READY/BUSY/PAUSED/SHUTDOWN). */
  private normalizeLaneStatus(s: unknown): string | undefined {
    if (typeof s !== 'string' || !s) return undefined;
    switch (s.toUpperCase()) {
      case 'IDLE':      return 'READY';
      case 'RUNNING':   return 'RUNNING';
      case 'COMPLETED': return 'SHUTDOWN';
      default:          return s.toUpperCase();
    }
  }

  /**
   * Called from the chat.tool_call / chat.tool_update branches of onJiuwenMessage.
   * memberName may be undefined if the server does not include it in the payload.
   * args (the parsed tool arguments) lets the feed say what an action touches —
   * e.g. a file_path, a glob pattern, or the shell command.
   */
  applyToolCall(toolName: string, filePath: string | undefined, memberName: string | undefined, args?: Record<string, unknown>): void {
    if (!memberName) return;
    const lane = this.getOrStubLane(memberName);
    lane.lastToolName = toolName;
    const activity = this.formatActivity(toolName, filePath, args);
    lane.currentActivity = activity;
    lane.lastActivePath = filePath ?? null;
    lane.lastActiveAt = Date.now();
    lane.status = 'BUSY';
    this.pushFeed(lane, activity, lane.lastActiveAt);
    this.lastEventAt = lane.lastActiveAt;
  }

  applyModelCallStart(memberName?: string): void {
    if (!memberName) return;
    const lane = this.getOrStubLane(memberName);
    lane.status = 'BUSY';
    lane.currentActivity = 'thinking…';
    lane.lastActiveAt = Date.now();
    this.pushFeed(lane, 'thinking…', lane.lastActiveAt);
    this.lastEventAt = lane.lastActiveAt;
  }

  /** First streamed token for a call: flip the lane from "thinking…" to "generating…". */
  applyModelTokenStart(memberName?: string): void {
    if (!memberName) return;
    const lane = this.lanes.get(memberName);
    if (!lane || lane.currentActivity !== 'thinking…') return;
    lane.currentActivity = 'generating…';
    lane.lastActiveAt = Date.now();
  }

  applyModelCallEnd(memberName?: string): void {
    if (!memberName) return;
    const lane = this.lanes.get(memberName);
    if (!lane) return;
    if (lane.currentActivity === 'thinking…' || lane.currentActivity === 'generating…') {
      lane.currentActivity = null;
    }
    lane.lastActiveAt = Date.now();
  }

  snapshot(): SwarmSnapshot {
    // Stable insertion order (join order) — lanes must not jump around as
    // statuses change; the webview renders status separately from position.
    const lanes = [...this.lanes.values()];
    const sortedTasks = [...this.tasks.values()].sort((a, b) => a.createdAt - b.createdAt);
    return {
      sessionId: this.sessionId,
      teamName: this.teamName,
      lanes,
      tasks: sortedTasks,
      messages: [...this.messages],
      lastEventAt: this.lastEventAt,
    };
  }

  reset(newSessionId: string): void {
    this.sessionId = newSessionId;
    this.teamName = '';
    this.lanes.clear();
    this.tasks.clear();
    this.messages = [];
    this.lastEventAt = 0;
  }

  isEmpty(): boolean {
    return this.lanes.size === 0;
  }

  // ──────────────────────────────────────────
  // Event handlers
  // ──────────────────────────────────────────

  private onMemberSpawned(e: Record<string, unknown>): void {
    const name = e['member_name'] as string | undefined;
    if (!name) return;
    const status = this.normalizeLaneStatus(e['status']) ?? 'READY';
    const prompt = e['prompt'] as string | undefined;
    const existing = this.lanes.get(name);
    if (existing) {
      existing.displayName = (e['display_name'] as string) ?? existing.displayName;
      existing.role = (e['role'] as string) ?? existing.role;
      existing.status = status;
      if (prompt) {
        existing.currentTaskTitle = prompt;
        existing.currentActivity = 'starting';
      }
      existing.startedAt = existing.startedAt ?? this.lastEventAt;
      this.pushFeed(existing, 'spawned', this.lastEventAt);
      existing.lastActiveAt = this.lastEventAt;
    } else {
      this.lanes.set(name, {
        memberName: name,
        displayName: (e['display_name'] as string) ?? name,
        role: (e['role'] as string) ?? 'TEAMMATE',
        status: status,
        executionStatus: 'IDLE',
        currentTaskId: null,
        currentTaskTitle: prompt ?? null,
        currentActivity: prompt ? 'starting' : null,
        lastToolName: null,
        lastActivePath: null,
        lastActiveAt: this.lastEventAt,
        startedAt: this.lastEventAt,
        activityFeed: [{ text: 'spawned', at: this.lastEventAt }],
        messageCount: 0,
        tasksDone: 0,
      });
    }
  }

  private onMemberStatusChanged(e: Record<string, unknown>): void {
    const name = e['member_name'] as string | undefined;
    if (!name) return;
    const lane = this.getOrStubLane(name);
    const status = this.normalizeLaneStatus(e['new_status'] ?? e['status']);
    if (status) lane.status = status;
    const outcome = e['outcome'] as string | undefined;
    if (outcome && status === 'SHUTDOWN') {
      lane.currentActivity = `done · ${outcome}`;
    }
    if (status === 'SHUTDOWN') this.pushFeed(lane, 'finished', this.lastEventAt);
    else if (status === 'PAUSED') this.pushFeed(lane, 'paused', this.lastEventAt);
    lane.lastActiveAt = this.lastEventAt;
  }

  private onMemberExecutionChanged(e: Record<string, unknown>): void {
    const name = e['member_name'] as string | undefined;
    if (!name) return;
    const lane = this.getOrStubLane(name);
    const execution = e['new_status'] ?? e['execution_status'];
    if (typeof execution === 'string' && execution) lane.executionStatus = execution.toUpperCase();
    lane.lastActiveAt = this.lastEventAt;
  }

  /** Live mid-run activity (e.g. a swarmflow worker tool call) → lane activity. */
  private onMemberActivityChanged(e: Record<string, unknown>): void {
    const name = e['member_name'] as string | undefined;
    if (!name) return;
    const activity = e['activity'] as string | undefined;
    if (!activity) return;
    const lane = this.getOrStubLane(name);
    lane.status = 'BUSY';
    lane.currentActivity = activity;
    this.pushFeed(lane, activity, this.lastEventAt);
    lane.lastActiveAt = this.lastEventAt;
  }

  private onMemberShutdown(e: Record<string, unknown>): void {
    const name = e['member_name'] as string | undefined;
    if (!name) return;
    const lane = this.lanes.get(name);
    if (!lane) return;
    lane.status = 'SHUTDOWN';
    lane.executionStatus = 'IDLE';
    lane.currentActivity = null;
    this.pushFeed(lane, 'shutdown', this.lastEventAt);
    lane.lastActiveAt = this.lastEventAt;
  }

  private onTaskCreated(e: Record<string, unknown>): void {
    const id = e['task_id'] as string | undefined;
    if (!id) return;
    this.tasks.set(id, {
      taskId: id,
      title: (e['title'] as string) ?? id,
      status: 'pending',
      assignee: null,
      createdAt: this.lastEventAt,
    });
  }

  private onTaskClaimed(e: Record<string, unknown>): void {
    const id = e['task_id'] as string | undefined;
    if (!id) return;
    const task = this.tasks.get(id)
      ?? { taskId: id, title: (e['title'] as string) ?? id, status: 'pending', assignee: null, createdAt: this.lastEventAt };
    this.tasks.set(id, task);
    // Gateway names the claimer member_id; normalizeEvent maps it to member_name.
    task.assignee = (e['assignee'] as string | null) ?? (e['member_name'] as string | null) ?? null;
    // Server carries the authoritative status (e.g. "in_progress") — prefer it;
    // fall back to started semantics for older gateways that only mark claimed.
    task.status = (e['status'] as string) ?? 'in_progress';
    // Link task into the lane so the board reflects the current work.
    if (task.assignee) {
      const lane = this.lanes.get(task.assignee);
      if (lane) {
        lane.currentTaskId = id;
        lane.currentTaskTitle = task.title;
        lane.lastActiveAt = this.lastEventAt;
      }
    }
  }

  /** Convergence event: apply a task's authoritative status/assignee snapshot. */
  private onTaskStatusSnapshot(e: Record<string, unknown>): void {
    const id = e['task_id'] as string | undefined;
    if (!id) return;
    const status = e['status'] as string | undefined;
    const assignee = (e['assignee'] as string | null) ?? (e['member_name'] as string | null) ?? null;
    const task = this.tasks.get(id)
      ?? { taskId: id, title: (e['title'] as string) ?? id, status: 'pending', assignee: null, createdAt: this.lastEventAt };
    if (status) task.status = status;
    if (assignee) task.assignee = assignee;
    this.tasks.set(id, task);
    if (task.assignee) {
      const lane = this.lanes.get(task.assignee);
      if (lane && task.status === 'completed' && lane.currentTaskId === id) {
        lane.currentTaskId = null;
        lane.currentTaskTitle = null;
        lane.tasksDone++;
        lane.lastActiveAt = this.lastEventAt;
      }
    }
  }

  private onTaskStarted(e: Record<string, unknown>): void {
    const id = e['task_id'] as string | undefined;
    if (!id) return;
    const task = this.tasks.get(id) ?? { taskId: id, title: id, status: 'pending', assignee: null, createdAt: this.lastEventAt };
    this.tasks.set(id, task);
    task.status = 'in_progress';
    task.assignee = (e['assignee'] as string) ?? task.assignee;
    if (task.assignee) {
      const lane = this.lanes.get(task.assignee);
      if (lane) {
        lane.currentTaskId = id;
        lane.currentTaskTitle = task.title;
        lane.lastActiveAt = this.lastEventAt;
      }
    }
  }

  private onTaskCompleted(e: Record<string, unknown>): void {
    const id = e['task_id'] as string | undefined;
    if (!id) return;
    const task = this.tasks.get(id);
    if (!task) return;
    task.status = 'completed';
    if (task.assignee) {
      const lane = this.lanes.get(task.assignee);
      if (lane && lane.currentTaskId === id) {
        lane.currentTaskId = null;
        lane.currentTaskTitle = null;
        lane.tasksDone++;
        lane.lastActiveAt = this.lastEventAt;
      }
    }
  }

  private onTaskCancelled(e: Record<string, unknown>): void {
    const id = e['task_id'] as string | undefined;
    if (!id) return;
    const task = this.tasks.get(id);
    if (!task) return;
    task.status = 'cancelled';
    if (task.assignee) {
      const lane = this.lanes.get(task.assignee);
      if (lane && lane.currentTaskId === id) {
        lane.currentTaskId = null;
        lane.currentTaskTitle = null;
        lane.lastActiveAt = this.lastEventAt;
      }
    }
  }

  private onTeamMessage(e: Record<string, unknown>): void {
    const from = e['from_member'] as string | undefined;
    if (!from) return;
    const lane = this.getOrStubLane(from);
    lane.messageCount++;
    lane.lastActiveAt = this.lastEventAt;
    // Capture message content for the inter-agent message log
    const content = e['content'] as string | undefined;
    if (content) {
      const to = (e['to_member'] as string | undefined) ?? null;
      this.messages.push({ from, to, content, timestamp: this.lastEventAt });
      if (this.messages.length > 50) this.messages.shift();
    }
  }

  // ──────────────────────────────────────────
  // Helpers
  // ──────────────────────────────────────────

  private getOrStubLane(memberName: string): AgentLane {
    const existing = this.lanes.get(memberName);
    if (existing) return existing;
    const now = this.lastEventAt || Date.now();
    const stub: AgentLane = {
      memberName,
      displayName: memberName,
      role: 'TEAMMATE',
      status: 'READY',
      executionStatus: 'IDLE',
      currentTaskId: null,
      currentTaskTitle: null,
      currentActivity: null,
      lastToolName: null,
      lastActivePath: null,
      lastActiveAt: now,
      startedAt: now,
      activityFeed: [],
      messageCount: 0,
      tasksDone: 0,
    };
    this.lanes.set(memberName, stub);
    return stub;
  }

  /** Append one entry to a lane's activity feed, keeping it capped.
   *  A duplicate of the previous entry within 1s (the tool_call + tool_update
   *  pair for one call) is collapsed into the existing row instead of flooding
   *  the feed; genuinely separate calls still get their own rows. */
  private pushFeed(lane: AgentLane, text: string, at: number): void {
    const last = lane.activityFeed[lane.activityFeed.length - 1];
    if (last && last.text === text && at - last.at < 1000) {
      last.at = at;
      return;
    }
    lane.activityFeed.push({ text, at });
    if (lane.activityFeed.length > 10) lane.activityFeed.shift();
  }

  private formatActivity(toolName: string, filePath: string | undefined, args?: Record<string, unknown>): string {
    const base = filePath ? path.basename(filePath) : undefined;
    const pattern = args && typeof args['pattern'] === 'string' ? args['pattern'] : undefined;
    const command = args && (
      (typeof args['command'] === 'string' && args['command']) ||
      (typeof args['cmd'] === 'string' && args['cmd']) ||
      (typeof args['command_line'] === 'string' && args['command_line'])
    );
    switch (toolName) {
      case 'write_file':
      case 'create_file':        return base ? `writing · ${base}` : 'writing';
      case 'str_replace_editor':
      case 'edit_file':          return base ? `editing · ${base}` : 'editing';
      case 'read_file':          return base ? `reading · ${base}` : 'reading';
      case 'bash':
      case 'run_command':
      case 'powershell':
      case 'shell':
      case 'cmd':                return command ? `running · ${this.trunc(command, 40)}` : 'running command';
      case 'list_files':         return pattern ? `listing · ${pattern}` : base ? `listing · ${base}` : 'listing';
      case 'glob':
      case 'grep':
      case 'search_files':       return pattern ? `searching · ${pattern}` : base ? `searching · ${base}` : 'searching';
      case 'view_task':          return args && args['task_id'] ? `viewing task · ${args['task_id']}` : 'viewing task';
      case 'claim_task':         return args && args['task_id'] ? `claiming task · ${args['task_id']}` : 'claiming task';
      case 'send_message':       return args && args['to'] ? `messaging · ${args['to']}` : 'messaging';
      default:                   return toolName;
    }
  }

  private trunc(s: string, n: number): string {
    return s.length > n ? s.slice(0, n) + '\u2026' : s;
  }
}
